builder/builder.go

```diff
@@ -0,0 +1,457 @@
+	}
+	pending := sync.Pending{
+		Block:       &pendingBlock,
+		StateUpdate: &su,
+		NewClasses:  newClasses,
+	}
+	b.pendingBlock.Store(&pending)
+	b.headState, b.headCloser, err = b.bc.HeadState()
+	return err
+}
+
+// Finalise the pending block and initialise the next one
+func (b *Builder) Finalise(signFunc blockchain.BlockSignFunc) error {
+	pending, err := b.Pending()
+	if err != nil {
+		return err
+	}
+	if err := b.bc.Finalise(pending.Block, pending.StateUpdate, pending.NewClasses, b.Sign); err != nil {
+		return err
+	}
+	b.log.Infow("Finalised block", "number", pending.Block.Number, "hash",
+		pending.Block.Hash.ShortString(), "state", pending.Block.GlobalStateRoot.ShortString())
+
+	if b.plugin != nil {
+		err := b.plugin.NewBlock(pending.Block, pending.StateUpdate, pending.NewClasses)
+		if err != nil {
+			b.log.Errorw("error sending new block to plugin", err)
+		}
+	}
+	// push the new block head to the feed
+	b.subNewHeads.Send(b.PendingBlock())
+	if err := b.ClearPending(); err != nil {
+		return err
+	}
+	return b.InitPendingBlock()
+}
+
+func (b *Builder) StartingBlockNumber() (uint64, error) {
+	return 0, nil
+}
+
+func (b *Builder) HighestBlockHeader() *core.Header {
+	return nil
+}
+
+// Sign returns the builder's signature over data.
+func (b *Builder) Sign(blockHash, stateDiffCommitment *felt.Felt) ([]*felt.Felt, error) {
+	data := crypto.PoseidonArray(blockHash, stateDiffCommitment).Bytes()
+	signatureBytes, err := b.privKey.Sign(data[:], nil)
+	if err != nil {
+		return nil, err
+	}
+	sig := make([]*felt.Felt, 0)
+	for start := 0; start < len(signatureBytes); {
+		step := len(signatureBytes[start:])
+		if step > felt.Bytes {
+			step = felt.Bytes
+		}
+		sig = append(sig, new(felt.Felt).SetBytes(signatureBytes[start:step]))
+		start += step
+	}
+	return sig, nil
+}
+
+// listenPool waits until the mempool has transactions, then
+// executes them one by one until the mempool is empty.
+func (b *Builder) listenPool(ctx context.Context) error {
+	for {
+		if err := b.depletePool(ctx); err != nil {
+			if !errors.Is(err, mempool.ErrTxnPoolEmpty) {
+				return err
+			}
+		}
+
+		// push the pending block to the feed
+		b.subPendingBlock.Send(b.PendingBlock())
+		select {
+		case <-ctx.Done():
+			return nil
+		// We wait for the mempool to get more txns before we continue
+		case <-b.mempool.Wait():
+			continue
+		}
+	}
+}
+
+// depletePool pops all available transactions from the mempool,
+// and executes them in sequence, applying the state changes
+// to the pending state
+func (b *Builder) depletePool(ctx context.Context) error {
+	blockHashToBeRevealed, err := b.getRevealedBlockHash()
+	if err != nil {
+		return err
+	}
+	for {
+		b.finaliseMutex.RLock()
+		userTxn, err := b.mempool.Pop()
+		if err != nil {
+			b.finaliseMutex.RUnlock()
+			return err
+		}
+		b.log.Debugw("running txn", "hash", userTxn.Transaction.Hash().String())
+		if err = b.runTxn(&userTxn, blockHashToBeRevealed); err != nil {
```

Comment:
I think you should send everything that's available on the mempool right now, no need to pop one by one